Mono.Debugger.Soft/Mono.Debugger.Soft/VirtualMachine.cs

```diff
@@ -177,30 +177,32 @@ internal void AddThreadToInvalidateList (ThreadMirror threadMirror)
 		}
 
 		public IList<ThreadMirror> GetThreads () {
-			var threads = threadCache;
-			if (threads == null) {
-				long[] ids = null;
-				var fetchingEvent = new ManualResetEvent (false);
-				vm.conn.VM_GetThreads ((threadsIds) => {
-					ids = threadsIds;
-					threadCache = threads = new ThreadMirror [threadsIds.Length];
-					fetchingEvent.Set ();
-				});
-				if (WaitHandle.WaitAny (new []{ vm.conn.DisconnectedEvent, fetchingEvent }) == 0) {
-					throw new VMDisconnectedException ();
+			lock (threadCacheLocker) {
+				var threads = threadCache;
+				if (threads == null) {
+					long[] ids = null;
+					var fetchingEvent = new ManualResetEvent (false);
+					vm.conn.VM_GetThreads ((threadsIds) => {
+						ids = threadsIds;
+						threadCache = threads = new ThreadMirror [threadsIds.Length];
+						fetchingEvent.Set ();
+					});
+					if (WaitHandle.WaitAny (new []{ vm.conn.DisconnectedEvent, fetchingEvent }) == 0) {
+						throw new VMDisconnectedException ();
+					}
+					for (int i = 0; i < ids.Length; ++i)
+						threads [i] = GetThread (ids [i]);
+					//Uncomment lines below if you want to re-fetch threads if new threads were started/stopped while
```

Comment:
And after this loop, set `threadCache=threads;`

Reply:
Yes yes, this is another possible solution that i discussed with @jstedfast, but we were afraid to call much more times the VM_GetThreads than it's really necessary. 
But with @lambdageek analysis, I think we prefer it than have a deadlock again.
I will do this.